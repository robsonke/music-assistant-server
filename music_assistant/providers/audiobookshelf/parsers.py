"""Parser for ABS -> MASS."""

from contextlib import suppress
from datetime import datetime

from aioaudiobookshelf.schema.library import (
    LibraryItemExpandedBook as AbsLibraryItemExpandedBook,
)
from aioaudiobookshelf.schema.library import (
    LibraryItemExpandedPodcast as AbsLibraryItemExpandedPodcast,
)
from aioaudiobookshelf.schema.library import (
    LibraryItemMinifiedBook as AbsLibraryItemMinifiedBook,
)
from aioaudiobookshelf.schema.library import (
    LibraryItemMinifiedPodcast as AbsLibraryItemMinifiedPodcast,
)
from aioaudiobookshelf.schema.library import (
    LibraryItemPodcast as AbsLibraryItemPodcast,
)
from aioaudiobookshelf.schema.media_progress import MediaProgress as AbsMediaProgress
from aioaudiobookshelf.schema.podcast import PodcastEpisodeExpanded as AbsPodcastEpisodeExpanded
from music_assistant_models.enums import ContentType, ImageType, MediaType
from music_assistant_models.media_items import Audiobook as MassAudiobook
from music_assistant_models.media_items import (
    AudioFormat,
    ItemMapping,
    MediaItemChapter,
    MediaItemImage,
    ProviderMapping,
    UniqueList,
)
from music_assistant_models.media_items import Podcast as MassPodcast
from music_assistant_models.media_items import PodcastEpisode as MassPodcastEpisode


def parse_podcast(
    *,
    abs_podcast: AbsLibraryItemExpandedPodcast
    | AbsLibraryItemMinifiedPodcast
    | AbsLibraryItemPodcast,
    lookup_key: str,
    domain: str,
    instance_id: str,
    token: str | None,
    base_url: str,
) -> MassPodcast:
    """Translate ABSPodcast to MassPodcast."""
    title = abs_podcast.media.metadata.title
    # Per API doc title may be None.
    if title is None:
        title = "UNKNOWN"
    mass_podcast = MassPodcast(
        item_id=abs_podcast.id_,
        name=title,
        publisher=abs_podcast.media.metadata.author,
        provider=lookup_key,
        provider_mappings={
            ProviderMapping(
                item_id=abs_podcast.id_,
                provider_domain=domain,
                provider_instance=instance_id,
            )
        },
    )
    mass_podcast.metadata.description = abs_podcast.media.metadata.description
    if token is not None:
        image_url = f"{base_url}/api/items/{abs_podcast.id_}/cover?token={token}"
        mass_podcast.metadata.images = UniqueList(
            [MediaItemImage(type=ImageType.THUMB, path=image_url, provider=lookup_key)]
        )
    mass_podcast.metadata.explicit = abs_podcast.media.metadata.explicit
    if abs_podcast.media.metadata.language is not None:
        mass_podcast.metadata.languages = UniqueList([abs_podcast.media.metadata.language])
    if abs_podcast.media.metadata.genres is not None:
        mass_podcast.metadata.genres = set(abs_podcast.media.metadata.genres)

    # podcast object has no published_at int, but an iso string
    if abs_podcast.media.metadata.release_date is not None:
        with suppress(ValueError):
            mass_podcast.metadata.release_date = datetime.fromisoformat(
                abs_podcast.media.metadata.release_date
            )

    if isinstance(abs_podcast, AbsLibraryItemExpandedPodcast | AbsLibraryItemPodcast):
        mass_podcast.total_episodes = len(abs_podcast.media.episodes)
    elif isinstance(abs_podcast, AbsLibraryItemMinifiedPodcast):
        mass_podcast.total_episodes = abs_podcast.media.num_episodes

    return mass_podcast


def parse_podcast_episode(
    *,
    episode: AbsPodcastEpisodeExpanded,
    prov_podcast_id: str,
    fallback_episode_cnt: int | None = None,
    lookup_key: str,
    domain: str,
    instance_id: str,
    token: str | None,
    base_url: str,
    media_progress: AbsMediaProgress | None = None,
) -> MassPodcastEpisode:
    """Translate ABSPodcastEpisode to MassPodcastEpisode.

    For an episode the id is set to f"{podcast_id} {episode_id}".
    ABS ids have no spaces, so we can split at a space to retrieve both
    in other functions.
    """
    url = f"{base_url}{episode.audio_track.content_url}"
    episode_id = f"{prov_podcast_id} {episode.id_}"

    release_date: datetime | None = None
    if episode.published_at is not None:
        position = -episode.published_at
        # abs published_at is ms epoch
        release_date = datetime.fromtimestamp(episode.published_at / 1000)
    else:
        position = 0
        if fallback_episode_cnt is not None:
            position = fallback_episode_cnt
    mass_episode = MassPodcastEpisode(
        item_id=episode_id,
        provider=lookup_key,
        name=episode.title,
        duration=int(episode.duration),
        position=position,
        podcast=ItemMapping(
            item_id=prov_podcast_id,
            provider=lookup_key,
            name=episode.title,
            media_type=MediaType.PODCAST,
        ),
        provider_mappings={
            ProviderMapping(
                item_id=episode_id,
                provider_domain=domain,
                provider_instance=instance_id,
                audio_format=AudioFormat(
                    content_type=ContentType.UNKNOWN,
                ),
                url=url,
            )
        },
    )

    mass_episode.metadata.release_date = release_date

    # cover image
    if token is not None:
        url_api = f"/api/items/{prov_podcast_id}/cover?token={token}"
        url_cover = f"{base_url}{url_api}"
        mass_episode.metadata.images = UniqueList(
            [MediaItemImage(type=ImageType.THUMB, path=url_cover, provider=lookup_key)]
        )

    if media_progress is not None and media_progress.current_time is not None:
        mass_episode.resume_position_ms = int(media_progress.current_time * 1000)
        mass_episode.fully_played = media_progress.is_finished

    return mass_episode


def parse_audiobook(
    *,
    abs_audiobook: AbsLibraryItemExpandedBook | AbsLibraryItemMinifiedBook,
    lookup_key: str,
    domain: str,
    instance_id: str,
    token: str | None,
    base_url: str,
    media_progress: AbsMediaProgress | None = None,
) -> MassAudiobook:
    """Translate AbsBook to Mass Book."""
    title = abs_audiobook.media.metadata.title
    # Per API doc title may be None.
    if title is None:
        title = "UNKNOWN TITLE"
    subtitle = abs_audiobook.media.metadata.subtitle
    if subtitle is not None or subtitle:
        title += f" | {subtitle}"
    mass_audiobook = MassAudiobook(
        item_id=abs_audiobook.id_,
        provider=lookup_key,
        name=title,
        duration=int(abs_audiobook.media.duration),
        provider_mappings={
            ProviderMapping(
                item_id=abs_audiobook.id_,
                provider_domain=domain,
                provider_instance=instance_id,
            )
        },
        publisher=abs_audiobook.media.metadata.publisher,
    )
    mass_audiobook.metadata.description = abs_audiobook.media.metadata.description
    if abs_audiobook.media.metadata.language is not None:
        mass_audiobook.metadata.languages = UniqueList([abs_audiobook.media.metadata.language])

    if abs_audiobook.media.metadata.published_date is not None:
        with suppress(ValueError):
            mass_audiobook.metadata.release_date = datetime.fromisoformat(
                abs_audiobook.media.metadata.published_date
            )
    elif abs_audiobook.media.metadata.published_year is not None:
        with suppress(ValueError):
            # ruff: noqa: DTZ001 # ignore tzinfo, this is a fallback attempt
            mass_audiobook.metadata.release_date = datetime(
                year=int(abs_audiobook.media.metadata.published_year), month=1, day=1
            )

    if abs_audiobook.media.metadata.genres is not None:
        mass_audiobook.metadata.genres = set(abs_audiobook.media.metadata.genres)

    mass_audiobook.metadata.explicit = abs_audiobook.media.metadata.explicit

    # cover
    if token is not None:
        api_url = f"/api/items/{abs_audiobook.id_}/cover?token={token}"
        cover_url = f"{base_url}{api_url}"
        mass_audiobook.metadata.images = UniqueList(
            [MediaItemImage(type=ImageType.THUMB, path=cover_url, provider=lookup_key)]
        )

    # expanded version
    if isinstance(abs_audiobook, AbsLibraryItemExpandedBook):
        mass_audiobook.authors.set([x.name for x in abs_audiobook.media.metadata.authors])
        mass_audiobook.narrators.set(abs_audiobook.media.metadata.narrators)
        chapters = []
        for idx, chapter in enumerate(abs_audiobook.media.chapters, 1):
            chapters.append(
                MediaItemChapter(
                    position=idx,
                    name=chapter.title,
                    start=chapter.start,
                    end=chapter.end,
                )
            )
        mass_audiobook.metadata.chapters = chapters

    elif isinstance(abs_audiobook, AbsLibraryItemMinifiedBook):
        mass_audiobook.authors.set([abs_audiobook.media.metadata.author_name])
        mass_audiobook.narrators.set([abs_audiobook.media.metadata.narrator_name])

    if media_progress is not None and media_progress.current_time is not None:
        mass_audiobook.resume_position_ms = int(media_progress.current_time * 1000)
        mass_audiobook.fully_played = media_progress.is_finished

    return mass_audiobook
