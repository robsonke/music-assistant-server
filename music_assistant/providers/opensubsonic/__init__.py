"""Open Subsonic music provider support for MusicAssistant."""

from __future__ import annotations

from typing import TYPE_CHECKING

from music_assistant_models.config_entries import ConfigEntry, ConfigValueType, ProviderConfig
from music_assistant_models.enums import ConfigEntryType

from music_assistant.constants import CONF_PASSWORD, CONF_PATH, CONF_PORT, CONF_USERNAME

from .sonic_provider import (
    CONF_BASE_URL,
    CONF_ENABLE_LEGACY_AUTH,
    CONF_ENABLE_PODCASTS,
    CONF_OVERRIDE_OFFSET,
    OpenSonicProvider,
)

if TYPE_CHECKING:
    from music_assistant_models.provider import ProviderManifest

    from music_assistant.mass import MusicAssistant
    from music_assistant.models import ProviderInstanceType


async def setup(
    mass: MusicAssistant, manifest: ProviderManifest, config: ProviderConfig
) -> ProviderInstanceType:
    """Initialize provider(instance) with given configuration."""
    return OpenSonicProvider(mass, manifest, config)


async def get_config_entries(
    mass: MusicAssistant,  # noqa: ARG001
    instance_id: str | None = None,  # noqa: ARG001
    action: str | None = None,  # noqa: ARG001
    values: dict[str, ConfigValueType] | None = None,  # noqa: ARG001
) -> tuple[ConfigEntry, ...]:
    """Return Config entries to setup this provider."""
    return (
        ConfigEntry(
            key=CONF_USERNAME,
            type=ConfigEntryType.STRING,
            label="Username",
            required=True,
            description="Your username for this Open Subsonic server",
        ),
        ConfigEntry(
            key=CONF_PASSWORD,
            type=ConfigEntryType.SECURE_STRING,
            label="Password",
            required=True,
            description="The password associated with the username",
        ),
        ConfigEntry(
            key=CONF_BASE_URL,
            type=ConfigEntryType.STRING,
            label="Base URL",
            required=True,
            description="Base URL for the server, e.g. https://subsonic.mydomain.tld",
        ),
        ConfigEntry(
            key=CONF_PORT,
            type=ConfigEntryType.INTEGER,
            label="Port",
            required=False,
            description="Port Number for the server",
        ),
        ConfigEntry(
            key=CONF_PATH,
            type=ConfigEntryType.STRING,
            label="Server Path",
            required=False,
            description="Path to append to the base URL for the Subsonic server, this is likely "
            "empty unless you are path routing on a proxy",
        ),
        ConfigEntry(
            key=CONF_ENABLE_PODCASTS,
            type=ConfigEntryType.BOOLEAN,
            label="Enable Podcasts",
            required=True,
            description="Should the provider query for podcasts as well as music?",
            default_value=True,
        ),
        ConfigEntry(
            key=CONF_ENABLE_LEGACY_AUTH,
            type=ConfigEntryType.BOOLEAN,
            label="Enable legacy auth",
            required=True,
            description='Enable OpenSubsonic "legacy" auth support',
            default_value=False,
        ),
        ConfigEntry(
            key=CONF_OVERRIDE_OFFSET,
            type=ConfigEntryType.BOOLEAN,
            label="Force player provider seek",
            required=True,
            description="Some Subsonic implementations advertise that they support seeking when "
            "they do not always. If seeking does not work for you, enable this.",
            default_value=False,
        ),
    )
