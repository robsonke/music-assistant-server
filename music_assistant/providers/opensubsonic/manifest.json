{
  "type": "music",
  "domain": "opensubsonic",
  "name": "Open Subsonic Media Server Library",
  "description": "Support for Open Subsonic based streaming providers in Music Assistant.",
  "codeowners": [
    "@khers"
  ],
  "requirements": [
    "py-opensonic==7.0.1"
  ],
  "documentation": "https://music-assistant.io/music-providers/subsonic/",
  "multi_instance": true
}
