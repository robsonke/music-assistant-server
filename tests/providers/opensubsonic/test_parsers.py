"""Test we can parse Open Subsonic models into Music Assistant models."""

import logging
import pathlib

import aiofiles
import pytest
from libopensonic.media import (
    Album,
    AlbumInfo,
    Artist,
    ArtistInfo,
    Playlist,
    PodcastChannel,
    PodcastEpisode,
)
from syrupy.assertion import SnapshotAssertion

from music_assistant.providers.opensubsonic.parsers import (
    parse_album,
    parse_artist,
    parse_epsiode,
    parse_playlist,
    parse_podcast,
)

FIXTURES_DIR = pathlib.Path(__file__).parent / "fixtures"
ARTIST_FIXTURES = list(FIXTURES_DIR.glob("artists/*.artist.json"))
ALBUM_FIXTURES = list(FIXTURES_DIR.glob("albums/*.album.json"))
PLAYLIST_FIXTURES = list(FIXTURES_DIR.glob("playlists/*.playlist.json"))
PODCAST_FIXTURES = list(FIXTURES_DIR.glob("podcasts/*.podcast.json"))
EPISODE_FIXTURES = list(FIXTURES_DIR.glob("episodes/*.episode.json"))

_LOGGER = logging.getLogger(__name__)


@pytest.mark.parametrize("example", ARTIST_FIXTURES, ids=lambda val: str(val.stem))
async def test_parse_artists(example: pathlib.Path, snapshot: SnapshotAssertion) -> None:
    """Test we can parse artists."""
    async with aiofiles.open(example) as fp:
        artist = Artist.from_json(await fp.read())

    parsed = parse_artist("xx-instance-id-xx", artist).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"].sort()
    assert snapshot == parsed

    # Find the corresponding info file
    example_info = example.with_suffix("").with_suffix(".info.json")
    async with aiofiles.open(example_info) as fp:
        artist_info = ArtistInfo.from_json(await fp.read())

    parsed = parse_artist("xx-instance-id-xx", artist, artist_info).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"].sort()
    assert snapshot == parsed


@pytest.mark.parametrize("example", ALBUM_FIXTURES, ids=lambda val: str(val.stem))
async def test_parse_albums(example: pathlib.Path, snapshot: SnapshotAssertion) -> None:
    """Test we can parse albums."""
    async with aiofiles.open(example) as fp:
        album = Album.from_json(await fp.read())

    parsed = parse_album(_LOGGER, "xx-instance-id-xx", album).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"].sort()
    assert snapshot == parsed

    # Find the corresponding info file
    example_info = example.with_suffix("").with_suffix(".info.json")
    async with aiofiles.open(example_info) as fp:
        album_info = AlbumInfo.from_json(await fp.read())

    parsed = parse_album(_LOGGER, "xx-instance-id-xx", album, album_info).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"].sort()
    assert snapshot == parsed


@pytest.mark.parametrize("example", PLAYLIST_FIXTURES, ids=lambda val: str(val.stem))
async def test_parse_playlist(example: pathlib.Path, snapshot: SnapshotAssertion) -> None:
    """Test we can parse Playlists."""
    async with aiofiles.open(example) as fp:
        playlist = Playlist.from_json(await fp.read())

    parsed = parse_playlist("xx-instance-id-xx", playlist).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"].sort()
    assert snapshot == parsed


@pytest.mark.parametrize("example", PODCAST_FIXTURES, ids=lambda val: str(val.stem))
async def test_parse_podcast(example: pathlib.Path, snapshot: SnapshotAssertion) -> None:
    """Test we can parse Podcasts."""
    async with aiofiles.open(example) as fp:
        podcast = PodcastChannel.from_json(await fp.read())

    parsed = parse_podcast("xx-instance-id-xx", podcast).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"].sort()
    assert snapshot == parsed


@pytest.mark.parametrize("example", EPISODE_FIXTURES, ids=lambda val: str(val.stem))
async def test_parse_episode(example: pathlib.Path, snapshot: SnapshotAssertion) -> None:
    """Test we can parse Podcast Episodes."""
    async with aiofiles.open(example) as fp:
        episode = PodcastEpisode.from_json(await fp.read())

    example_channel = example.with_suffix("").with_suffix(".podcast.json")
    async with aiofiles.open(example_channel) as fp:
        channel = PodcastChannel.from_json(await fp.read())

    parsed = parse_epsiode("xx-instance-id-xx", episode, channel).to_dict()
    # sort external Ids to ensure they are always in the same order for snapshot testing
    parsed["external_ids"].sort()
    assert snapshot == parsed
