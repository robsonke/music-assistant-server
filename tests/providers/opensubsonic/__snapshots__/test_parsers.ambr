# serializer version: 1
# name: test_parse_albums[spec.album]
  dict({
    'album_type': 'unknown',
    'artists': list([
      dict({
        'available': True,
        'external_ids': list([
        ]),
        'image': None,
        'is_playable': True,
        'item_id': '91c3901ac465b9efc439e4be4270c2b6',
        'media_type': 'artist',
        'name': 'pornophonique',
        'provider': 'xx-instance-id-xx',
        'sort_name': 'pornophonique',
        'translation_key': None,
        'uri': 'xx-instance-id-xx://artist/91c3901ac465b9efc439e4be4270c2b6',
        'version': '',
      }),
    ]),
    'external_ids': list([
    ]),
    'favorite': True,
    'is_playable': True,
    'item_id': 'ad0f112b6dcf83de5e9cae85d07f0d35',
    'media_type': 'album',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'al-ad0f112b6dcf83de5e9cae85d07f0d35_640a93a8',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'languages': None,
      'last_refresh': None,
      'links': None,
      'lrc_lyrics': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'release_date': None,
      'review': None,
      'style': None,
    }),
    'name': '8-bit lagerfeuer',
    'position': None,
    'provider': 'opensubsonic',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 0,
          'channels': 2,
          'codec_type': '?',
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': 'ad0f112b6dcf83de5e9cae85d07f0d35',
        'provider_domain': 'opensubsonic',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': '8-bit lagerfeuer',
    'translation_key': None,
    'uri': 'opensubsonic://album/ad0f112b6dcf83de5e9cae85d07f0d35',
    'version': '',
    'year': 2007,
  })
# ---
# name: test_parse_albums[spec.album].1
  dict({
    'album_type': 'unknown',
    'artists': list([
      dict({
        'available': True,
        'external_ids': list([
        ]),
        'image': None,
        'is_playable': True,
        'item_id': '91c3901ac465b9efc439e4be4270c2b6',
        'media_type': 'artist',
        'name': 'pornophonique',
        'provider': 'xx-instance-id-xx',
        'sort_name': 'pornophonique',
        'translation_key': None,
        'uri': 'xx-instance-id-xx://artist/91c3901ac465b9efc439e4be4270c2b6',
        'version': '',
      }),
    ]),
    'external_ids': list([
    ]),
    'favorite': True,
    'is_playable': True,
    'item_id': 'ad0f112b6dcf83de5e9cae85d07f0d35',
    'media_type': 'album',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': 'Download the full release here (creative commons). These cripsy beats are ripe with thumping funk and techno influences, sample wizardry and daring shuffles. Composed with the help of unique sound plugins which were especially programmed to measure Comfort Fit’s needs and wishes, we think the chances aren’t bad that you’ll fall for the unique sound signature, bounce and elegance of this unusual Hip Hop production. Ltj bukem / Good looking Rec., UK: "Really love this music." Velanche / XLR8R, UK: "Awesome job he\'s done... overall production is dope." Kwesi / BBE Music, UK: "Wooooooowwwww... WHAT THE FUCK! THIS IS WHAT',
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'al-ad0f112b6dcf83de5e9cae85d07f0d35_640a93a8',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
        dict({
          'path': 'http://localhost:8989/play/art/0f8c3cbd6b0b22c3b5402141351ac812/album/21/thumb34.jpg',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'languages': None,
      'last_refresh': None,
      'links': None,
      'lrc_lyrics': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'release_date': None,
      'review': None,
      'style': None,
    }),
    'name': '8-bit lagerfeuer',
    'position': None,
    'provider': 'opensubsonic',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 0,
          'channels': 2,
          'codec_type': '?',
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': 'ad0f112b6dcf83de5e9cae85d07f0d35',
        'provider_domain': 'opensubsonic',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': '8-bit lagerfeuer',
    'translation_key': None,
    'uri': 'opensubsonic://album/ad0f112b6dcf83de5e9cae85d07f0d35',
    'version': '',
    'year': 2007,
  })
# ---
# name: test_parse_artists[spec-artistid3.artist]
  dict({
    'external_ids': list([
    ]),
    'favorite': True,
    'is_playable': True,
    'item_id': '37ec820ca7193e17040c98f7da7c4b51',
    'media_type': 'artist',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'ar-37ec820ca7193e17040c98f7da7c4b51_0',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'languages': None,
      'last_refresh': None,
      'links': None,
      'lrc_lyrics': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'release_date': None,
      'review': None,
      'style': None,
    }),
    'name': '2 Mello',
    'position': None,
    'provider': 'opensubsonic',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 0,
          'channels': 2,
          'codec_type': '?',
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '37ec820ca7193e17040c98f7da7c4b51',
        'provider_domain': 'opensubsonic',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': '2 mello',
    'translation_key': None,
    'uri': 'opensubsonic://artist/37ec820ca7193e17040c98f7da7c4b51',
    'version': '',
  })
# ---
# name: test_parse_artists[spec-artistid3.artist].1
  dict({
    'external_ids': list([
    ]),
    'favorite': True,
    'is_playable': True,
    'item_id': '37ec820ca7193e17040c98f7da7c4b51',
    'media_type': 'artist',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': 'Empty biography',
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'ar-37ec820ca7193e17040c98f7da7c4b51_0',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
        dict({
          'path': 'http://localhost:8989/play/art/f20070e8e11611cc53542a38801d60fa/artist/2/thumb34.jpg',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': True,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'languages': None,
      'last_refresh': None,
      'links': None,
      'lrc_lyrics': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'release_date': None,
      'review': None,
      'style': None,
    }),
    'name': '2 Mello',
    'position': None,
    'provider': 'opensubsonic',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 0,
          'channels': 2,
          'codec_type': '?',
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '37ec820ca7193e17040c98f7da7c4b51',
        'provider_domain': 'opensubsonic',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': '2 mello',
    'translation_key': None,
    'uri': 'opensubsonic://artist/37ec820ca7193e17040c98f7da7c4b51',
    'version': '',
  })
# ---
# name: test_parse_artists[spec-sample.artist]
  dict({
    'external_ids': list([
    ]),
    'favorite': True,
    'is_playable': True,
    'item_id': '100000002',
    'media_type': 'artist',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': None,
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'ar-100000002',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'languages': None,
      'last_refresh': None,
      'links': None,
      'lrc_lyrics': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'release_date': None,
      'review': None,
      'style': None,
    }),
    'name': 'Synthetic',
    'position': None,
    'provider': 'opensubsonic',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 0,
          'channels': 2,
          'codec_type': '?',
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '100000002',
        'provider_domain': 'opensubsonic',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': 'synthetic',
    'translation_key': None,
    'uri': 'opensubsonic://artist/100000002',
    'version': '',
  })
# ---
# name: test_parse_artists[spec-sample.artist].1
  dict({
    'external_ids': list([
    ]),
    'favorite': True,
    'is_playable': True,
    'item_id': '100000002',
    'media_type': 'artist',
    'metadata': dict({
      'chapters': None,
      'copyright': None,
      'description': 'Empty biography',
      'explicit': None,
      'genres': None,
      'images': list([
        dict({
          'path': 'ar-100000002',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': False,
          'type': 'thumb',
        }),
        dict({
          'path': 'http://localhost:8989/play/art/f20070e8e11611cc53542a38801d60fa/artist/2/thumb34.jpg',
          'provider': 'xx-instance-id-xx',
          'remotely_accessible': True,
          'type': 'thumb',
        }),
      ]),
      'label': None,
      'languages': None,
      'last_refresh': None,
      'links': None,
      'lrc_lyrics': None,
      'lyrics': None,
      'mood': None,
      'performers': None,
      'popularity': None,
      'preview': None,
      'release_date': None,
      'review': None,
      'style': None,
    }),
    'name': 'Synthetic',
    'position': None,
    'provider': 'opensubsonic',
    'provider_mappings': list([
      dict({
        'audio_format': dict({
          'bit_depth': 16,
          'bit_rate': 0,
          'channels': 2,
          'codec_type': '?',
          'content_type': '?',
          'output_format_str': '?',
          'sample_rate': 44100,
        }),
        'available': True,
        'details': None,
        'item_id': '100000002',
        'provider_domain': 'opensubsonic',
        'provider_instance': 'xx-instance-id-xx',
        'url': None,
      }),
    ]),
    'sort_name': 'synthetic',
    'translation_key': None,
    'uri': 'opensubsonic://artist/100000002',
    'version': '',
  })
# ---
